engine/audiomixer.cpp

```diff
@@ -0,0 +1,167 @@
+    return true;
+}
+
+void AudioMixer::Cleanup()
+{
+    PaError err = Pa_StopStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream stopping failed: {}", Pa_GetErrorText(err));
+        return;
+    }
+
+    err = Pa_CloseStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream closing failed: {}", Pa_GetErrorText(err));
+        return;
+    }
+
+    Pa_Terminate();
+}
+
+void AudioMixer::PlaySound(const Sound& sound)
+{
+    mPlayingSounds.emplace_back(sound);
+}
+
+void AudioMixer::PlaySound(const Sound& sound, const glm::vec3& position)
+{
+    mPlayingSounds.emplace_back(sound, position);
+}
+
+int AudioMixer::AudioCallback(const void* /*inputBuffer*/, void* outputBuffer,
+                              unsigned long framesPerBuffer,
+                              const PaStreamCallbackTimeInfo* /*timeInfo*/,
+                              PaStreamCallbackFlags /*statusFlags*/,
+                              void* /*userData*/)
+{
+    float* out = static_cast<float*>(outputBuffer);
+
+    int frames = static_cast<int>(framesPerBuffer);
+    for (int i = 0; i < frames; i++)
+    {
+        float sampleL = 0.0f;
+        float sampleR = 0.0f;
+
+        for (auto& sound : mPlayingSounds)
+        {
+            if (sound.IsFinished())
+            {
+                continue;
+            }
+
+            if (sound.GetNumChannels() >= 2)
+            {
+                sampleL += sound.GetSample(0);
+                sampleR += sound.GetSample(1);
+                sound.NextSample();
+            }
+            else
+            {
+                sampleL += sound.GetSample(0);
+                sampleR += sound.GetSample(0);
+                sound.NextSample();
+            }
+
+            if (mSpatialAudioEnabled && sound.IsPositional())
+            {
+                const glm::vec3 soundPosition = sound.GetPosition();
+                const float leftDistance = glm::distance(LEFT_LISTENER_POSITION, soundPosition);
+                const float rightDistance = glm::distance(RIGHT_LISTENER_POSITION, soundPosition);
+
+                // calculate attenuation
+                float leftAttenuation = 1.0f / (leftDistance / MIN_DISTANCE);
+                float rightAttenuation = 1.0f / (rightDistance / MIN_DISTANCE);
+
+                leftAttenuation = std::clamp(leftAttenuation, 0.0f, 1.0f);
+                rightAttenuation = std::clamp(rightAttenuation, 0.0f, 1.0f);
+
+                sampleL *= leftAttenuation;
+                sampleR *= rightAttenuation;
+            }
+        }
+
+        sampleL *= mVolume;
+        sampleR *= mVolume;
+
+        sampleL = std::clamp(sampleL, -1.0f, 1.0f);
+        sampleR = std::clamp(sampleR, -1.0f, 1.0f);
+
+        *out++ = sampleL;
+        *out++ = sampleR;
+    }
+
+    mPlayingSounds.erase(std::remove_if(mPlayingSounds.begin(), mPlayingSounds.end(), [](const auto& sound) {
+        return sound.IsFinished();
+    }), mPlayingSounds.end());
```

Comment:
Not realtime safe, unfortunately :( (also not thread safe)

again, I would consider skipping this for now as the fix is going to be complex and painful :) yay audio

Reply:
my fav 😁